keps/sig-api-machinery/4988-serve-pagination-from-cache/README.md

```diff
@@ -0,0 +1,410 @@
+  - Looks up the "nextSmaller" snapshot based on the resourceVersion.
+  - Constructs the response using data from the retrieved snapshot.
+  **Edge cases:**
+    - Requested resourceVersion is smaller than any available snapshot:
+      This indicates that the requested data has been cleaned up. In this scenario,
+      the API server falls back to serving the request from etcd.
+    - Requested resourceVersion is larger than the latest snapshot:
+      This could indicate a future resourceVersion or a situation where the watch
+      cache is lagging behind. The API server performs a consistent read from
+      etcd to confirm the existence of the future resourceVersion or waits for
+      the watch cache to catch up.
+4. **Snapshot Cleanup:** Snapshots are subject to a Time-To-Live (TTL) mechanism
+   similar to watch events. The proposed approach leverages the existing process
+   that limits the number of events to 10,000 within a 75-second window
+   (configurable via request timeout). Additionally, snapshots are purged during
+   cache re-initialization.
+
+[Clone()]: https://pkg.go.dev/github.com/google/btree#BTree.Clone
+
+### Cache Inconsistency detection
+
+For the first iteration of the mechanism we will calculate the hash based on LIST response
+at some specific RV of each resource and expose it as a metric. Hash will be calculated for
+both cache and etcd, allowing users to setup an alert to detect if they don't match.
+
+Once every 5 minutes (default compaction interval) we will send a LIST request to both etcd and
+watch cache and calculate hash of the response.
+
+TODO: Decide whether to calculate hash based on object fields or values stored in etcd.
+
+### Risks and Mitigations
+
+#### Memory overhead
+
+B-tree snapshots are designed to minimize memory overhead by storing pointers to
+the actual objects, rather than the objects themselves. Since the objects are
+already cached to serve watch events, the primary memory impact comes from the
+B-tree structure itself. To quantify the memory overhead, we run 5k scalability tests.
+They should represent the worst case scenario, as they utilize large number of small objects.
+The results are promising:
+
+* **Object Allocations:** Allocation profile collected during the test test has
+  shown an increase of 7GB in object allocations, which translates to a
+  negligible 0.2% of total allocations.
+* **Memory Usage:** Memory in use profile collected during the test has shown
+  Btree memory usage of 300MB, representing a 1.3% of total memory used.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+- Ensure the pagination is well tested
+
+##### Unit tests
+
+- `k8s/apiserver/pkg/storage/cache`: `2024-12-12` - `<test coverage>`
+
+##### Integration tests
```

Comment:
I'd like an integration test covering the "fallback to etcd on mismatch" case to be sure it works well.

Reply:
I was thinking more about purging the cache, but agree we need a test for it. Added the proposal for a test, we can discuss what action we should  in above comment.